core/chains/solana/chain.go

```diff
@@ -44,13 +46,136 @@ type chain struct {
 	lggr           logger.Logger
 
 	// tracking node chain id for verification
-	clientCache map[string]cachedClient // map URL -> {client, chainId} [mainnet/testnet/devnet/localnet]
+	clientCache map[string]*verifiedCachedClient // map URL -> {client, chainId} [mainnet/testnet/devnet/localnet]
 	clientLock  sync.RWMutex
 }
 
-type cachedClient struct {
-	id string
-	rw solanaclient.ReaderWriter
+type verifiedCachedClient struct {
+	chainID         string
+	expectedChainID string
+	nodeURL         string
+
+	lggr logger.Logger
+
+	verifyErr           error
+	chainIDVerified     bool
+	chainIDVerifiedLock sync.RWMutex
+
+	solanaclient.ReaderWriter
+}
+
+func (v *verifiedCachedClient) verifyChainID() bool {
+	v.chainIDVerifiedLock.RLock()
+	if v.chainIDVerified {
+		return true
+	}
+	v.chainIDVerifiedLock.RUnlock()
+
+	var err error
+
+	v.chainIDVerifiedLock.Lock()
+	defer v.chainIDVerifiedLock.Unlock()
```

Comment:
Can a chain ever become unverified? If so you'll need to re-check inside this mutex.

Reply:
It cannot become unverified, the check only happens on chain creation for the chain id, afterwards it is considered verified, as it was in the previous implementation.

In short, the only status change that will happen is unverified -> verified.